### DISCUSSION TITLE
Grails Profiles for Grails 6 and 7

### DISCUSSION THREAD

**Discussion by jamesfredley** [apache/grails-core#13583] [Polls]
Grails Shell, Wrapper and Profiles can optionally be used in Grails 6 to bridge the following gaps and provide time for them to be migrated to gradle, directly:

- IntelliJ Grails Plugin - requires shell which requires profiles
- grailsw in the current project using wrapper
- custom company/internal grails scripts
- grails scripts/commands that have not been migrated to gradle such as `grails list-plugins` and `grails dbm-update`
- custom company/internal profiles

[Grails Profiles](https://github.com/grails-profiles) have not been updated since early in Grails 5, but "work" on Grails 6.

Grails Shell 6.1.2 was released with Grails 6.1.2 and Grails Shell 6.2.1 is soon to be released with Grails 6.2.1.

[Grails Wrapper 4.0.0](https://github.com/grails/grails-wrapper/releases/tag/v4.0.0) for Grails 6 is available and there is a PR to add it as an optional feature on Grails Forge

Example Grails 6.2 app with Grails Shell, Wrapper and Web Profile:  https://github.com/jamesfredley/Grails62ShellWrapperProfilesDemo
diff to add Grails Shell, Wrapper and Web Profile: https://github.com/jamesfredley/Grails62ShellWrapperProfilesDemo/commit/72c4fef786fc9d1462288fdba46a7320485c9b90

**How do you use Grails Profiles?**

**Comment by TheRetroRoot** [comment]
The removal of profiles has stopped us from upgrading to 6 at the moment. We have a few internal profile we use that have been rendered useless. Also, not sure if it is 100% related but I don't see value in using the Grails forge. Much rather just have the shell as it always has been.

Let's keep the framework friendly and not invalidate years of documentation by removing the simple `grails` command.

**Reply by jamesfredley** [reply]
The Grails Forge CLI introduction has created confusion.  Over the last few weeks I've reviewing the changes and worked to make sure we can continue to use Grails Shell, Wrapper and Profiles to fill current gaps.

I added a quick Grails 6.2 demo app with Grails Shell, Wrapper and Web Profile to the poll description above.

**Comment by gsartori** [comment]
We had a profile to startup Dueuno Elements applications, it would be great to make it usable again.

**Reply by jamesfredley** [reply]
How far do the following changes get you and what is missing/broken?

https://github.com/jamesfredley/Grails62ShellWrapperProfilesDemo/commit/72c4fef786fc9d1462288fdba46a7320485c9b90

**Reply by gsartori** [reply]
Dueuno Elements is just a dependency to be added to a standard Grails project so a new Dueuno Elements project by itself does not require the profile dependencies. What we had was a Grails profile implementation we used to create new apps from the command line since we have some manual operations to be done on each new project.

**Comment by jamesfredley** [comment]
@TheRetroRoot & @gsartori : I've published initial 6.0.0-SNAPSHOT's of Grails Web and Base Profiles which are compatible with Grails 6.  The skeleton(contents) are the same as the 5 release with the exception of Grails 6 compatible Grails Wrapper and Gradle 7.6.4 wrapper files, so this is a starting place to add additional changes necessary on the skeleton and dependencies.

[jamesfredley/Grails62ShellWrapperProfilesDemo](https://github.com/jamesfredley/Grails62ShellWrapperProfilesDemo) - can be tested with `./grailsw` or IntelliJ Grails Plugin run configuration.

**Reply by jamesfredley** [reply]
To test create-app with these initial 6.0.0-SNAPSHOT's of Grails Web and Base Profiles

On the 6.2.x branch of [grails-core](https://github.com/grails/grails-core)

In `grails-bom/profiles.properties`
Change

```
base=6.0.0-SNAPSHOT
web=6.0.0-SNAPSHOT
```

Run `assemble` on grails-core

grab grails-6.2.1-SNAPSHOT.zip from build/distributions

run the `grails `command

**Reply by jamesfredley** [reply]
All of the profiles now have initial new Grails 6 compatible snapshots.

```
rest-api=6.0.0-SNAPSHOT
base=6.0.0-SNAPSHOT
web=6.0.0-SNAPSHOT
react=6.0.0-SNAPSHOT
vue=6.0.0-SNAPSHOT
plugin=6.0.0-SNAPSHOT
profile=6.0.0-SNAPSHOT
rest-api-plugin=6.0.0-SNAPSHOT
web-plugin=6.0.0-SNAPSHOT
angular=9.0.0-SNAPSHOT
```

**Reply by jamesfredley** [reply]
web-plugin 6.0.0-SNAPSHOT has now been published

**Reply by rainboyan** [reply]
@jamesfredley Glad to see that, you did a good job!

**Reply by jamesfredley** [reply]
Grails 7 compatible initial Profile snapshots have been created.

The skeleton(contents) are the same as the 5 release with the exception of Grails 7 compatible Grails Wrapper and Gradle 7.6.4 wrapper files, so this is a just a starting place to add additional changes necessary on the skeleton and dependencies.

```
rest-api=7.0.0-SNAPSHOT
base=7.0.0-SNAPSHOT
web=7.0.0-SNAPSHOT
react=7.0.0-SNAPSHOT
vue=7.0.0-SNAPSHOT
plugin=7.0.0-SNAPSHOT
profile=7.0.0-SNAPSHOT
rest-api-plugin=7.0.0-SNAPSHOT
web-plugin=7.0.0-SNAPSHOT
angular=10.0.0-SNAPSHOT
```

Grails 7 compatible grails-wrapper Snapshot:  https://github.com/grails/grails-wrapper/releases/tag/v5.0.0

Grails 7 Shell Snapshot:  `org.grails:grails-shell:7.0.0-SNAPSHOT`

**Comment by rainboyan** [comment]
I agree with @TheRetroRoot, Grails Shell is better than Grails Forge, not only for generating projects but also provides many commands, the most important thing is that it allows users to create private Profiles for internal projects.

The missing features are lack of support Groovy Markup Templates and custom profiles are not allowed to be published in Grails namespace, so they are not easily shared and discovered by the community.

I created a custom profile for better Bootstrap CSS support https://github.com/rainboyan/web-bootstrap/, only a few people know that, it's not easy to use custom profile in Grails.

I also implemented Application Templates for Grace, support Groovy template engine to generate files and use Groovy Ant tasks to write scripts and hooks, you can create your own templates or boilerplates and share them on GitHub, the other developers can use them too.
I'm glad to see that it works well for Grails 3/4/5/6.

**Comment by jamesfredley** [comment]
Grails Shell will be returning in Grails 6.2.1 and is in the current Grails 7.0.0-SNAPSHOT.

Grails Wrapper has been released for Grails 6 and 7 and it will be an optional feature in Grails Forge 6.2.1 and 7.0.0, using either the grails distribution or https://start.grails.org/.  Releases: https://github.com/grails/grails-wrapper/releases

And finally Grails 6 and Grails 7 compatible profiles have been published.  These exist to support Grails Shell CLI which requires Profiles and users with custom internal profiles.  They have not been fully upgraded to generate up to date Grails 6 and 7 apps via create-app.  Grails Forge, either the grails distribution or https://start.grails.org/ should be used for creating a new apps.

**Grails 6 profile versions:**
rest-api=6.0.0
base=6.0.0
web=6.0.0
react=6.0.0
vue=6.0.0
plugin=6.0.0
profile=6.0.0
rest-api-plugin=6.0.0
web-plugin=6.0.0
angular=9.0.0

**Grails 7 profile versions:**
rest-api=7.0.0
base=7.0.0
web=7.0.0
react=7.0.0
vue=7.0.0
plugin=7.0.0
profile=7.0.0
rest-api-plugin=7.0.0
web-plugin=7.0.0
angular=10.0.0
